tests/4.5/target_enter_exit_data/test_target_enter_exit_data_map_global_array.c

```diff
@@ -1,45 +1,88 @@
+//===--- test_target_enter_exit_data_map_global_array.c ---------------------===//
+//
+// OpenMP API Version 4.5 Nov 2015
+//
+// This test is in two parts. First, the test checks that mapping to on enter
+// followed by mapping from on exit works, by modifying the data on the
+// device. Then, the delete clause is tested by making sure that deleting
+// an array mapped to the device resets its reference count, meaning that
+// modifications made on the host are remapped back in when another map(to)
+// is encountered.
+//
+////===----------------------------------------------------------------------===//
+
 #include <stdio.h>
 #include <stdlib.h>
+#include "ompvv.h"
 #include <omp.h>
 
-// Test for OpenMP 4.5 target enter and target exit data with global arrays.
+#define N 10
 
-int n=10;
-int *x;
-int A[10]={0,0,0,0,0,0,0,0,0,0};
+int A[N] = {0};
+int B[N] = {0};
 
-void init(){
- #pragma omp target enter data map(to:A[:n])
-}
+int test_tofrom() {
+  int errors = 0;
 
-void finalize(){
- #pragma omp target exit data map(from:A[:n])
-}
+  for (int i = 0; i < N; ++i) {
+    A[i] = 0;
+  }
+
+#pragma omp target enter data map(to: A)
 
-int main (){
-int isHost=-1,i,errors=0;
+#pragma omp target
+  {
+    for (int i = 0; i < N; i++) {
+      A[i] = N;
+    }
+  }
 
-init();
+#pragma omp target exit data map(from: A)
 
-#pragma omp target map(tofrom: isHost)  
-{
- /*Record where the computation was executed*/
- isHost = omp_is_initial_device();
+  for (int i = 0; i < N; ++i) {
+    OMPVV_TEST_AND_SET_VERBOSE(errors, A[i] != N);
+  }
 
- for(i=0;i< n; i++)
-   A[i] = 10;
+  return errors;
 }
 
- finalize();
- for(i=0; i<n; i++)
-   if(A[i] != 10){
-     errors += 1;
-   }
+int test_delete() {
+  int errors = 0;
+
+  for (int i = 0; i < N; ++i) {
+    A[i] = N;
+  }
+
+#pragma omp target data map(tofrom: A) map(from: B)
+  {
+#pragma omp target exit data map(delete: A)
+    for (int i = 0; i < N; ++i) {
+      A[i] = 0;
+    }
+#pragma omp target map(to: A)   // if the delete does not work, this map will not happen.
```

Comment:
Is it not possible to map to A once the reference count becomes 0?

Reply:
Yup, unless you have the `always` clause. The map will only happen if the reference count is 1 or if the `always` clause is present.
Check out 4.5 spec page 217, lines 17-20.

Reply:
Thanks!